### ISSUE TITLE
Certain videos appear more saturated than in normal media player apps

### ISSUE THREAD

**Issue by frenzibyte** [ppy/osu-framework#6402]
On the right is o!f and on the left is QuickTime Player:

![CleanShot 2024-10-27 at 12 24 05](https://github.com/user-attachments/assets/4580ee42-626f-4365-9f1f-526994a6394a)

Video: https://discord.com/channels/90072389919997952/892258907135836212/1297071581230530611 (to be replaced with a sharable video later).

**Comment by Joehuu** [comment]
Probably related to https://github.com/ppy/osu/issues/30399.

**Comment by frenzibyte** [comment]
That does not seem to be relevant as far as I can see. In fact, I see zero difference between o!f and a media player on the existing test sample video `h264.mp4`.

**Comment by frenzibyte** [comment]
Investigations show that since the video is marked with `AVCOL_SPC_UNSPECIFIED` colorspace, it's looking odd in o!f because it's using these [matrix coefficients](https://github.com/ppy/osu-framework/blob/b4ed7217c1b2f66e27c96cc264a53fbde1cd7948/osu.Framework/Graphics/Video/VideoDecoder.cs#L281-L284), whereas forcing it to use the [matrix coefficients](https://github.com/ppy/osu-framework/blob/b4ed7217c1b2f66e27c96cc264a53fbde1cd7948/osu.Framework/Graphics/Video/VideoDecoder.cs#L274-L276) for Rec. 709 renders the video closer? to correct colours:

![CleanShot 2024-10-28 at 01 52 48](https://github.com/user-attachments/assets/1ac89a26-fce0-498b-adbc-4ae3f703c91c)

I'm still not entirely sure why this video in particular is faulty in o!f while its metadata almost exactly matches the sample video `h264.mp4`, which I can't notice anything wrong with it. The only thing I can find about this video is that it was exported with Premiere Pro.

**Comment by peppy** [comment]
Just change it to use rec709 if unspecified and call it a day.

But where are you seeing the "video is marked with"? Are you saying that this differs from the other test video? Your description is unclear.

**Comment by frenzibyte** [comment]
The other test video also has unspecified colorspace, and using the rec709 matrix for unspecified makes the test video look wrong (yellow/cyan/green colours looks more dimmed).

Test video with the SMPTE170M/SMPTE240M matrix coefficients (i.e. master):

![CleanShot 2024-10-28 at 02 09 40](https://github.com/user-attachments/assets/10da387c-37d2-4511-b1cb-d7cd2b80db63)

Test video with the BT709 matrix coefficients (above proposal):

![CleanShot 2024-10-28 at 02 11 45](https://github.com/user-attachments/assets/7d51966a-1694-4a29-8366-5c2217e9390e)

**Comment by peppy** [comment]
fwiw the owc video also looks different in quicktime vs iina (mpv/ffmpeg based):

![IINA 2024-10-28 at 06 26 51](https://github.com/user-attachments/assets/04a260c4-3338-4d8b-9b08-30c51876f9b3)

**Comment by frenzibyte** [comment]
The results are actually a bit weird, IINA is incorrect while bare mpv is correct;
![CleanShot 2024-10-28 at 23 17 52](https://github.com/user-attachments/assets/a660ec84-9b88-4414-9bd9-1af1ab31d4d9)

Furthermore, I was randomly searching through VLC's code and I stumbled across this:
https://github.com/videolan/vlc/blob/456d7c92575dd226c8b967d312d5a67b4fa8fb55/include/vlc_es.h#L458-L464

Then I compared it with QuickTime Player's behaviour by scaling the sample video to specific resolutions with ffmpeg while keeping the colorspace tag as `unspecified`, and it turns out QuickTime uses the Rec. 709 colorspace for any video with a width higher than 704 or height higher than 576. I'll push a PR inline with that behaviour seeing that both VLC and QuickTime Player do it.
